0.data_analysis/utils/CorrelateData.py

```diff
@@ -0,0 +1,174 @@
+
+        _posthoc_group_cols: List(Objects)
+            Columns to groupby after organizing by _antehoc_group_cols.
+
+        _feat_col: List
+            Feature columns.
+
+        _posthoc_group_cols: List(Objects)
+            Columns to groupby befor organizing by _posthoc_group_cols.
+
+        _drop_col: List
+            Columns to not save in the final output.
+
+        Returns
+        -------
+        corrs: pd.DataFrame
+            The correlated data including columns from both the ante and post hoc groups less the dropped columns.
+        """
+
+        groupdf = _df.groupby(_antehoc_group_cols)
+
+        # Retrieve keys for the ante group
+        gkeys = groupdf.groups.keys()
+
+        # Find all possible combinations of size 2 ante group keys
+        apairs = list(combinations(gkeys, 2))
+
+        # Store correlations
+        corrs = defaultdict(list)
+
+        # Iterate through each ante group combination
+        for apair in apairs:
+
+            # Extract the keys for the first post hoc group
+            group0df = groupdf.get_group(apair[0]).copy()
+            group0df = group0df.groupby(_posthoc_group_cols)[_feat_cols]
+            group0_keys = group0df.groups.keys()
+
+            # Extract the keys for the second post hoc group
+            group1df = groupdf.get_group(apair[1]).copy()
+            group1df = group1df.groupby(_posthoc_group_cols)[_feat_cols]
+            group1_keys = group1df.groups.keys()
+
+            # Iterate through each well group cartesian product and save the data
+            for ppair in list(product(group0_keys, group1_keys)):
+
+                # Store correlations
+                corrs["correlation"].append(group0df.get_group(ppair[0]).iloc[0].corr(
+                    group1df.get_group(ppair[1]).iloc[0]
+                ))
+
+                # Save the column data for both the ante and post hoc groups less the dropped columns
+                corrs = self.save_params(apair, _antehoc_group_cols, _drop_cols, corrs)
+                corrs = self.save_params(ppair, _posthoc_group_cols, _drop_cols, corrs)
+
+        return pd.DataFrame(corrs)
+
+
+    def intra_correlations(self, _df, _posthoc_group_cols, _feat_cols, _antehoc_group_cols, _drop_cols = []):
+        """
+        Computes correlations between two objects using both post hoc and ante groups.
+        This is accomplished by computing correlations between objects only in the same ante group (intra correlations), but between different post hoc groups.
+
+        Parameters
+        ----------
+        _df: pd.DataFrame
+            Contains the features and group columns to use for correlating objects.
+
+        _posthoc_group_cols: List(Objects)
+            Columns to groupby after organizing by _antehoc_group_cols.
+
+        _feat_col: List
+            Feature columns.
+
+        _posthoc_group_cols: List(Objects)
+            Columns to groupby befor organizing by _posthoc_group_cols.
```

Comment:
Same issue as above.
```suggestion
        _antehoc_group_cols: List(Objects)
            Columns to groupby before organizing by _posthoc_group_cols.
```

Reply:
Changed